File: backend/src/agents/tools/script-tools.ts
/**
 * 剧本改写 Agent 工具
 * 工厂函数模式 — 注入 episodeId，工具不再需要 LLM 传递 ID
 */
import { createTool } from '@mastra/core/tools'
import { z } from 'zod'
import { db, schema } from '../../db/index.js'
import { eq } from 'drizzle-orm'
import { now } from '../../utils/response.js'
import { sliceLongText } from '../../utils/text-slice.js'

export function createScriptTools(episodeId: number) {
  const readEpisodeScript = createTool({
    id: 'read_episode_script',
    description: 'Read the script content of the current episode.',
    inputSchema: z.object({}),
    execute: async () => {
      const [ep] = db.select().from(schema.episodes)
        .where(eq(schema.episodes.id, episodeId)).all()
      if (!ep) return { error: `Episode not found (id=${episodeId})` }
      const content = ep.content || ep.scriptContent
      if (!content) return { error: `Episode has no content (id=${episodeId})` }
      const sliced = sliceLongText(content)
      return {
        content: sliced.text,
        word_count: content.length,
        episode_id: episodeId,
        truncated: sliced.truncated,
        total_chars: sliced.total_chars,
      }
    },
  })

  const rewriteToScreenplay = createTool({
    id: 'rewrite_to_screenplay',
    description: 'Read the original content for AI rewriting. Returns the source text with formatting instructions.',
    inputSchema: z.object({
      instructions: z.string().optional().describe('Additional rewrite instructions'),
    }),
    execute: async ({ instructions }) => {
      const [ep] = db.select().from(schema.episodes)
        .where(eq(schema.episodes.id, episodeId)).all()
      if (!ep) return { error: `Episode not found` }
      const source = ep.content || ep.scriptContent
      if (!source) return { error: `Episode has no content to rewrite` }
      const sliced = sliceLongText(source)

      return {
        source_content: sliced.text,
        truncated: sliced.truncated,
        total_chars: sliced.total_chars,
        instruction: `请将以下内容改写为格式化剧本。

输出硬性要求：
- 只输出剧本正文，禁止输出任何分析、评论、解读、剧情预测或总结。
- 对白必须逐角色分行：每个角色说的话单独占一行，一人一句换行，禁止把两个及以上人物的对话写在同一段。
- 对白用第一人称直接引语，禁止第三人称转述（如"小雪说……""他说道……"）。

说话人划分铁律（谁说的话，最高优先级）：
- 每一句台词都必须清楚标注是谁说的，禁止多人台词混在一段。
- 一句一行、一行一人。
- 叙述体拆解归因：原文是叙述体（如"何长青叹了口气说……林雪抬头道……"）时，必须拆成两行各自归到正确角色名下，绝不能整段照抄。
- 归因准确：根据上下文判断说话人，不能张冠李戴。
- 示例：【错误】"何长青叹了口气，说自己也不知道。林雪轻声说别担心。"【正确】"何长青：（叹气）我也不知道。"换行"林雪：（轻声）别担心，我们一起想办法。"

格式规范：
- 场景头：## S编号 | 内景/外景 · 地点 | 时间段
- 动作描写：自然段落，不包含镜头语言
- 对白：角色名：（状态/表情）台词内容
- 每个场景 30-60 秒内容

${instructions || ''}

【原始内容】
${sliced.text}`,
      }
    },
  })

  const saveScript = createTool({
    id: 'save_script',
    description: 'Save the rewritten screenplay content to the current episode.',
    inputSchema: z.object({
      content: z.string().describe('The formatted screenplay content to save'),
    }),
    execute: async ({ content }) => {
      db.update(schema.episodes)
        .set({ scriptContent: content, updatedAt: now() })
        .where(eq(schema.episodes.id, episodeId))
        .run()
      // 剧本内容指纹门禁：改写落库后重算指纹
      await import('../../services/script-fingerprint.js').then(m => m.refreshEpisodeScriptHash(episodeId))
      return { message: `Script saved`, word_count: content.length }
    },
  })

  return { readEpisodeScript, rewriteToScreenplay, saveScript }
}

File: backend-py/app/services/agents/tools/script_tools.py
"""剧本改写工具（移植自 ``agents/tools/script-tools.ts``，83 行）。

三个工具组成「读 → 改写 → 存」闭环：

* ``read_episode_script``   读本集原文（**超长自动滑窗截断**，见 ``text_slice``）；
* ``rewrite_to_screenplay`` 读原文 **+ 注入格式化剧本规范**（``SCREENPLAY_FORMAT_RULES``）
  一并返回给模型 —— 规范只在 ``prompt_blocks`` 定义一次，杜绝副本漂移；
* ``save_script``           落库并**重算剧本指纹**（指纹门禁：改写后旧指纹必须失效）。

⚠️ 三处保真点：

* 原文取值链是 ``content || scriptContent``（两个字段都可能承载原文）；
* 三个工具的报错文案各不相同（``… (id=N)`` 带 id 的两条只在 ``read`` 里）；
* 保存后**必须**刷新指纹，否则下游（分镜/提取）会拿着过期指纹判断"剧本没变"。
"""
from __future__ import annotations

from typing import Any

from sqlalchemy import select, update
from sqlalchemy.engine import Connection

from ....db import engine
from ....models import episodes
from ....response import now
from ...prompt_blocks import SCREENPLAY_FORMAT_RULES
from ...script_fingerprint import refresh_episode_script_hash
from ...text_slice import slice_long_text
from ..tool import Tool, json_string, object_schema

__all__ = ["create_script_tools"]


def create_script_tools(episode_id: int) -> dict[str, Tool]:
    """工厂（闭包注入 ``episodeId``；每个工具自开短事务）。"""

    def _episode_source(conn: Connection) -> tuple[bool, str]:
        """返回 ``(是否找到剧集, 原文)``。"""
        row = conn.execute(
            select(episodes.c.content, episodes.c.script_content)
            .where(episodes.c.id == episode_id)
        ).first()
        if row is None:
            return False, ""
        return True, (row[0] or row[1] or "")

    async def read_episode_script(_arguments: dict[str, Any]) -> dict[str, Any]:
        with engine.begin() as conn:
            found, content = _episode_source(conn)
        if not found:
            return {"error": f"Episode not found (id={episode_id})"}
        if not content:
            return {"error": f"Episode has no content (id={episode_id})"}
        sliced = slice_long_text(content)
        return {
            "content": sliced["text"],
            "word_count": len(content),
            "episode_id": episode_id,
            "truncated": sliced["truncated"],
            "total_chars": sliced["total_chars"],
        }

    async def rewrite_to_screenplay(arguments: dict[str, Any]) -> dict[str, Any]:
        instructions = arguments.get("instructions")
        with engine.begin() as conn:
            found, source = _episode_source(conn)
        if not found:
            return {"error": "Episode not found"}
        if not source:
            return {"error": "Episode has no content to rewrite"}
        sliced = slice_long_text(source)
        return {
            "source_content": sliced["text"],
            "truncated": sliced["truncated"],
            "total_chars": sliced["total_chars"],
            "instruction": (
                "请将以下内容改写为格式化剧本。\n\n"
                f"{SCREENPLAY_FORMAT_RULES}\n\n"
                f"{instructions or ''}\n\n"
                "【原始内容】\n"
                f"{sliced['text']}"
            ),
        }

    async def save_script(arguments: dict[str, Any]) -> dict[str, Any]:
        content = arguments.get("content") or ""
        with engine.begin() as conn:
            conn.execute(
                update(episodes).where(episodes.c.id == episode_id).values(
                    script_content=content, updated_at=now()
                )
            )
            # 剧本内容指纹门禁：改写落库后重算指纹
            refresh_episode_script_hash(conn, episode_id)
        return {"message": "Script saved", "word_count": len(content)}

    return {
        "read_episode_script": Tool(
            id="read_episode_script",
            description="Read the script content of the current episode.",
            input_schema=object_schema({}, required=[]),
            execute=read_episode_script,
        ),
        "rewrite_to_screenplay": Tool(
            id="rewrite_to_screenplay",
            description=(
                "Read the original content for AI rewriting. "
                "Returns the source text with formatting instructions."
            ),
            input_schema=object_schema(
                {"instructions": json_string("Additional rewrite instructions")}, required=[]
            ),
            execute=rewrite_to_screenplay,
        ),
        "save_script": Tool(
            id="save_script",
            description="Save the rewritten screenplay content to the current episode.",
            input_schema=object_schema(
                {"content": json_string("The formatted screenplay content to save")}
            ),
            execute=save_script,
        ),
    }
